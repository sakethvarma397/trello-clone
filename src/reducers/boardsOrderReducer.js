import { ADD_BOARD, DELETE_BOARD } from "../utils/constants";
const boardsOrderReducer = (state = ["board-1", "board-2"], action) => {
  switch (action.type) {
    case ADD_BOARD: {
      const { id } = action.payload;
      const newState = state.slice();
      newState.push(`board-${id}`);

      return newState;
    }

    case DELETE_BOARD: {
      const { boardId } = action.payload;
      const newState = state.filter((id) => id !== boardId);
      return newState;
    }

    default:
      return state;
  }
};

export default boardsOrderReducer;
