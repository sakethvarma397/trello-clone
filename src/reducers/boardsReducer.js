import {
  ADD_BOARD,
  ADD_LIST,
  DELETE_BOARD,
  DELETE_LIST,
  EDIT_BOARD_TITLE,
  ON_DRAG,
} from "../utils/constants";
const initialState = {
  "board-1": {
    id: "board-1",
    title: "First board",
    lists: ["list-1"],
  },
  "board-2": {
    id: "board-2",
    title: "Second board",
    lists: ["list-2"],
  },
};

const boardReducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_BOARD: {
      const { title, id } = action.payload;
      const newId = `board-${id}`;
      const newBoard = {
        id: newId,
        title,
        lists: [],
      };

      return { ...state, [newId]: newBoard };
    }

    case EDIT_BOARD_TITLE: {
      const { boardId, title } = action.payload;
      const board = state[boardId];
      board.title = title;

      return { ...state, [boardId]: board };
    }

    case DELETE_BOARD: {
      const { boardId } = action.payload;
      const newState = { ...state };
      delete newState[boardId];
      return newState;
    }

    case ADD_LIST: {
      const { boardId, id } = action.payload;
      const board = state[boardId];
      const newListId = `list-${id}`;
      const newLists = [...board.lists, newListId];
      board.lists = newLists;
      return { ...state, [boardId]: board };
    }

    case DELETE_LIST: {
      const { listId, boardId } = action.payload;
      const board = state[boardId];
      const lists = board.lists;
      const newLists = lists.filter((id) => id !== listId);
      board.lists = newLists;
      return { ...state, [boardId]: board };
    }

    case ON_DRAG: {
      const { boardId } = action.payload;
      const board = state[boardId];
      const lists = board.lists;
      const {
        droppableIndexEnd,
        droppableIndexStart,

        type,
      } = action.payload;

      // Only handles the drag between lists
      if (type === "list") {
        const pulledOutList = lists.splice(droppableIndexStart, 1);
        lists.splice(droppableIndexEnd, 0, ...pulledOutList);
        board.lists = lists;

        return { ...state, [boardId]: board };
      }
      return state;
    }

    default:
      return state;
  }
};

export default boardReducer;
