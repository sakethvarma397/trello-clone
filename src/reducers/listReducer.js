import { ADD_LIST, ADD_TASK, ON_DRAG } from "../utils/constants";
let task_Id = 5;
let list_Id = 3;
const initialState = {
  lists: [
    {
      id: "list-1",
      title: "Hello",
      tasks: [
        { id: "task-1", text: "First card", listId: "list-1" },
        { id: "task-2", text: "Second Card", listId: "list-1" },
      ],
    },
    {
      id: "list-2",
      title: "There",
      tasks: [
        { id: "task-3", text: "First card", listId: "list-2" },
        { id: "task-4", text: "Second Card", listId: "list-2" },
      ],
    },
  ],
};

export const listReducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_LIST: {
      const newList = {
        id: `list-${list_Id}}`,
        tasks: [],
        title: action.title,
      };
      list_Id += 1;
      return { lists: [...state.lists, newList] };
    }
    case ADD_TASK: {
      const { listId, text } = action.details;
      const newTask = {
        text,
        listId,
        id: `task-${task_Id}`,
      };
      task_Id += 1;
      const newState = state.lists.map((list) => {
        if (list.id === newTask.listId) {
          return { ...list, tasks: [...list.tasks, newTask] };
        } else {
          return list;
        }
      });
      return { lists: newState };
    }
    case ON_DRAG: {
      const {
        droppableIdStart,
        droppableIdEnd,
        droppableIndexStart,
        droppableIndexEnd,
        type,
      } = action.payload;
      const newLists = state.lists.slice();

      if (type === "list") {
        const list = newLists.splice(droppableIndexStart, 1);
        newLists.splice(droppableIndexEnd, 0, ...list);
      }

      if (droppableIdStart === droppableIdEnd && type !== "list") {
        const list = newLists.find((list) => list.id === droppableIdStart);
        const task = list.tasks.splice(droppableIndexStart, 1);
        list.tasks.splice(droppableIndexEnd, 0, ...task);
      }
      if (droppableIdStart !== droppableIdEnd) {
        const sourceList = newLists.find(
          (list) => list.id === droppableIdStart
        );
        const destlist = newLists.find((list) => list.id === droppableIdEnd);
        const task = sourceList.tasks.splice(droppableIndexStart, 1);
        destlist.tasks.splice(droppableIndexEnd, 0, ...task);
      }

      return { lists: newLists };
    }
    default:
      return state;
  }
};
