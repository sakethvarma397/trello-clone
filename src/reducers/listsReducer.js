import {
  ADD_LIST,
  ADD_TASK,
  DELETE_TASK,
  DELETE_LIST,
  EDIT_LIST_TITLE,
  ON_DRAG,
} from "../utils/constants";

const initialState = {
  "list-1": {
    id: "list-1",
    title: "Hello",
    tasks: ["task-1"],
  },
  "list-2": {
    id: "list-2",
    title: "There",
    tasks: ["task-2"],
  },
};
const listsReducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_LIST: {
      const { title, id } = action.payload;
      const newList = {
        title: title,
        id: `list-${id}`,
        tasks: [],
      };

      const newLists = { ...state, [`list-${id}`]: newList };

      return newLists;
    }

    case DELETE_LIST: {
      const { listId } = action.payload;
      const newState = { ...state };
      delete newState[listId];
      return newState;
    }

    case EDIT_LIST_TITLE: {
      const newLists = state;
      const { listId, listTitle } = action.payload;
      const list = newLists[listId];
      list.title = listTitle;
      return { ...state, [listId]: list };
    }

    case ADD_TASK: {
      const { listId, id } = action.payload;
      const list = { ...state[listId] };
      list.tasks.push(`task-${id}`);
      return { ...state, [listId]: list };
    }

    case ON_DRAG: {
      const {
        droppableIdStart,
        droppableIdEnd,
        droppableIndexStart,
        droppableIndexEnd,
        type,
      } = action.payload;

      //Changing list order wouldn't be handled here
      if (type !== "list") {
        // Moving tasks within a list
        if (droppableIdStart === droppableIdEnd) {
          const list = state[droppableIdStart];
          const task = list.tasks.splice(droppableIndexStart, 1);
          list.tasks.splice(droppableIndexEnd, 0, ...task);
          return { ...state, [droppableIdStart]: list };
        } else {
          // Moving tasks across the lists
          // first find the list where the drag happened
          const listStart = state[droppableIdStart];
          // pull out the card from this list
          const task = listStart.tasks.splice(droppableIndexStart, 1);
          // find the list where the drag ended
          const listEnd = state[droppableIdEnd];

          // put the card in the new list
          listEnd.tasks.splice(droppableIndexEnd, 0, ...task);
          return {
            ...state,
            [droppableIdStart]: listStart,
            [droppableIdEnd]: listEnd,
          };
        }
      }
      return { ...state };
    }

    case DELETE_TASK: {
      const { listId, id } = action.payload;
      const list = { ...state[listId] };
      const newTasks = list.tasks.filter((task) => task.id !== id);
      return { ...state, [listId]: { ...list, tasks: newTasks } };
    }

    default:
      return state;
  }
};

export default listsReducer;
