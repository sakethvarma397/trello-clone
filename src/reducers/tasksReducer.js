import {
  ADD_TASK,
  DELETE_TASK,
  EDIT_TASK,
  TOGGLE_TASK,
} from "../utils/constants";

const initialState = {
  "task-1": {
    text: "Sample text",
    id: `task-1`,
    list: "list-1",
    description: "",
    isPending: false,
  },
  "task-2": {
    text: "Sample text",
    id: `task-2`,
    list: "list-2",
    description: "",
    isPending: false,
  },
};
const taskReducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_TASK: {
      const { text, listId, id } = action.payload;
      const newTask = {
        text,
        id: `task-${id}`,
        list: listId,
        descrption: "",
        isPending: false,
      };

      return { ...state, [`task-${id}`]: newTask };
    }

    //TODO: Implement this
    case EDIT_TASK: {
      return state;
    }

    case DELETE_TASK: {
      const { id } = action.payload;
      const newState = state;
      delete newState[id];
      return newState;
    }

    case TOGGLE_TASK: {
      const { isPending, taskId } = action.payload;
      const task = state[taskId];
      task.isPending = isPending;
      return { ...state, [taskId]: task };
    }

    default:
      return state;
  }
};
export default taskReducer;
