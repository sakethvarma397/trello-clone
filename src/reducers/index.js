import { combineReducers } from "redux";
import listsReducer from "./listsReducer";
import tasksReducer from "./tasksReducer";
import boardsReducer from "./boardsReducer";
import boardsOrderReducer from "./boardsOrderReducer";
import storage from "redux-persist/lib/storage";
import { persistReducer } from "redux-persist";

const persistConfig = {
  key: "root",
  storage: storage,
  whitelist: ["lists", "tasks", "boards", "boardsOrder"],
};

const rootReducer = combineReducers({
  lists: listsReducer,
  tasks: tasksReducer,
  boards: boardsReducer,
  boardsOrder: boardsOrderReducer,
});

export default persistReducer(persistConfig, rootReducer);
