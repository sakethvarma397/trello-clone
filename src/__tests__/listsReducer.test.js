import listsReducer from "../reducers/listsReducer";
import { initialState } from "../reducers/listsReducer";
import {
  ADD_LIST,
  ADD_TASK,
  DELETE_LIST,
  EDIT_LIST_TITLE,
  DELETE_TASK,
} from "../utils/constants";
describe("Lists reducer", () => {
  it("should return the initial state", () => {
    expect(listsReducer(undefined, {})).toMatchSnapshot();
  });

  it("should handle DELETE_LIST", () => {
    expect(
      listsReducer(initialState, {
        type: DELETE_LIST,
        payload: {
          listId: "list-1",
        },
      })
    ).toMatchSnapshot();
  });

  it("should handle ADD_LIST", () => {
    expect(
      listsReducer(initialState, {
        type: ADD_LIST,
        payload: {
          title: "Hello",
          id: "12",
        },
      })
    ).toMatchSnapshot();
  });

  it("should handle EDIT_LIST_TITLE", () => {
    expect(
      listsReducer(initialState, {
        type: EDIT_LIST_TITLE,
        payload: {
          listId: "list-1",
          listTitle: "New Title",
        },
      })
    ).toMatchSnapshot();
  });

  it("should handle ADD_TASK", () => {
    expect(
      listsReducer(initialState, {
        type: ADD_TASK,
        payload: {
          listId: "list-1",
          id: "4",
        },
      })
    ).toMatchSnapshot();
  });

  it("should handle DELETE_TASK", () => {
    expect(
      listsReducer(initialState, {
        type: DELETE_TASK,
        payload: {
          listId: "list-1",
          id: "task-1",
        },
      })
    ).toMatchSnapshot();
  });
});
