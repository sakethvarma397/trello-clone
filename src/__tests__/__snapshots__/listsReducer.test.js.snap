// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lists reducer should handle ADD_LIST 1`] = `
Object {
  "list-1": Object {
    "id": "list-1",
    "tasks": Array [
      "task-1",
    ],
    "title": "Hello",
  },
  "list-12": Object {
    "id": "list-12",
    "tasks": Array [],
    "title": "Hello",
  },
  "list-2": Object {
    "id": "list-2",
    "tasks": Array [
      "task-2",
    ],
    "title": "There",
  },
}
`;

exports[`Lists reducer should handle ADD_TASK 1`] = `
Object {
  "list-1": Object {
    "id": "list-1",
    "tasks": Array [
      "task-1",
      "task-4",
    ],
    "title": "New Title",
  },
  "list-2": Object {
    "id": "list-2",
    "tasks": Array [
      "task-2",
    ],
    "title": "There",
  },
}
`;

exports[`Lists reducer should handle DELETE_LIST 1`] = `
Object {
  "list-2": Object {
    "id": "list-2",
    "tasks": Array [
      "task-2",
    ],
    "title": "There",
  },
}
`;

exports[`Lists reducer should handle DELETE_TASK 1`] = `
Object {
  "list-1": Object {
    "id": "list-1",
    "tasks": Array [
      "task-4",
    ],
    "title": "New Title",
  },
  "list-2": Object {
    "id": "list-2",
    "tasks": Array [
      "task-2",
    ],
    "title": "There",
  },
}
`;

exports[`Lists reducer should handle EDIT_LIST_TITLE 1`] = `
Object {
  "list-1": Object {
    "id": "list-1",
    "tasks": Array [
      "task-1",
    ],
    "title": "New Title",
  },
  "list-2": Object {
    "id": "list-2",
    "tasks": Array [
      "task-2",
    ],
    "title": "There",
  },
}
`;

exports[`Lists reducer should return the initial state 1`] = `
Object {
  "list-1": Object {
    "id": "list-1",
    "tasks": Array [
      "task-1",
    ],
    "title": "Hello",
  },
  "list-2": Object {
    "id": "list-2",
    "tasks": Array [
      "task-2",
    ],
    "title": "There",
  },
}
`;
